tests/ModelContextProtocol.Tests/Client/McpClientExtensionsTests.cs

```diff
@@ -376,4 +380,125 @@ public async Task WithDescription_ChangesToolDescription()
+            })]),
+            cancellationToken: token);
+
+        // Act
+        await client.NotifyCancelAsync(
+            requestId: new("abc"),
+            reason: "Cancelled",
+            cancellationToken: token);
+        await Assert.ThrowsAsync<TimeoutException>(
+            async () => await clientReceived.Task
+                .WaitAsync(TimeSpan.FromSeconds(5), token));
+        // Assert
+        Assert.False(clientReceived.Task.IsCompleted);
+    }
+
+    [Fact]
+    public async Task Can_Notify_Cancel()
+    {
+        // Arrange
+        var token = TestContext.Current.CancellationToken;
+        await using var client = await CreateMcpClientForServer(
+            cancellationToken: token);
+        RequestId expectedRequestId = new("abc");
+        var expectedReason = "Cancelled";
+
+        // Act
+        await client.NotifyCancelAsync(
+            requestId: expectedRequestId,
+            reason: expectedReason,
+            cancellationToken: token);
+
+        // Assert
+        var result = await _clientToServerPipe.Reader.ReadAsync(token);
+        var copyBytes = new byte[result.Buffer.Length];
+        result.Buffer.CopyTo(copyBytes);
+        Utf8JsonReader reader = new(copyBytes);
+        var jsonRpcNotification = JsonSerializer.Deserialize<JsonRpcNotification>(ref reader);
+        Assert.NotNull(jsonRpcNotification);
+        Assert.Equal(NotificationMethods.CancelledNotification, jsonRpcNotification.Method);
+
+        var parameters = jsonRpcNotification.Params;
+        Assert.NotNull(parameters);
+        var cancelledNotification = JsonSerializer.Deserialize<CancelledNotification>(parameters.ToString());
+        Assert.NotNull(cancelledNotification);
+        Assert.Equal(expectedRequestId.ToString(), cancelledNotification.RequestId.ToString());
+        Assert.Equal(expectedReason, cancelledNotification.Reason);
+    }
+
+    private McpClientOptions CreateClientOptions(
+        IEnumerable<KeyValuePair<string, Func<JsonRpcNotification, Task>>>? notificationHandlers = null)
+        => new()
+        {
+            Capabilities = new()
+            {
+                NotificationHandlers = notificationHandlers ?? [],
+            },
+        };
+
+    private async Task NotifyClientAsync(
+        string message, object? parameters = null, CancellationToken token = default)
+        => await NotifyPipeAsync(_serverToClientPipe, message, parameters, token);
+    private async static Task NotifyPipeAsync(
+        Pipe pipe, string message, object? parameters = null, CancellationToken token = default)
+    {
+        var bytes = JsonSerializer.SerializeToUtf8Bytes(new JsonRpcNotification
+        {
+            Method = message,
+            Params = parameters is not null ? JsonSerializer.Serialize(parameters) : null,
+        });
+        await pipe.Writer.WriteAsync(bytes, token);
+        await pipe.Writer.CompleteAsync(); // Signal the end of the message
```

Comment:
CompleteAsync isn't about the end of a message, it's about saying nothing more will ever be written to the pipe. Is that intended?